include/boost/math/quadrature/detail/tanh_sinh_detail.hpp

```diff
@@ -0,0 +1,689 @@
+            // but for more complicated integrands, this calculation is not noticeable.
+            Real abterm = (abs(yp) + abs(ym))*w;
+            absum += abterm;
+        }
+
+        I1 += sum*h;
+        L1_I1 += absum*h;
+        ++k;
+        err = abs(I0 - I1);
+        // std::cout << "Estimate:        " << I1 << " Error estimate at level " << k  << " = " << err << std::endl;
+
+        if (!(boost::math::isfinite)(I1))
+        {
+            return policies::raise_evaluation_error(function, "The tanh_sinh quadrature evaluated your function at a singular point at got %1%. Please narrow the bounds of integration or check your function for singularities.", I1, Policy());
+        }
+        //
+        // Termination condition:
+        // No more levels are considered once the error is less than the specified tolerance.
+        // Note however, that we always go down at least 4 levels, otherwise we risk missing
+        // features of interest in f() - imagine for example a function which flatlines, except
+        // for a very small "spike". An example would be the incomplete beta integral with large
+        // parameters.  We could keep hunting until we find something, but that would handicap
+        // integrals which really are zero.... so a compromise then!
+        //
+        if ((k > 4) && (err <= tolerance*L1_I1))
+        {
+            break;
+        }
+
+    }
+    if (error)
+    {
+        *error = err;
+    }
+
+    if (L1)
+    {
+        *L1 = L1_I1;
+    }
+
+    if (levels)
+    {
+       *levels = k;
+    }
+
+    return I1;
+}
+
+template<class Real, class Policy>
+void tanh_sinh_detail<Real, Policy>::init(const Real& min_complement, const mpl::int_<0>&)
+{
+   using std::tanh;
+   using std::sinh;
+   using std::asinh;
+   using std::atanh;
+   using std::ceil;
+   using boost::math::constants::half_pi;
+   using boost::math::constants::pi;
+   using boost::math::constants::two_div_pi;
+
+   m_committed_refinements = 4;
+   //
+   // Initial row length is one step to the right of the abscissa value
+   // that would go to the full extent of the requested range, we need this
+   // to ensure full precision as otherwise we chop off quite a chunk of the
+   // range in *subsequent* rows.
+   //
+   m_inital_row_length = itrunc(ceil(t_from_abscissa_complement(min_complement)));
+   std::size_t first_complement = 0;
+   m_t_max = m_inital_row_length;
+   m_t_crossover = t_from_abscissa_complement(Real(0.5f));
+
+   m_abscissas.assign(m_max_refinements + 1, std::vector<Real>());
+   m_weights.assign(m_max_refinements + 1, std::vector<Real>());
+   m_first_complements.assign(m_max_refinements + 1, 0);
+   //
+   // First row is special:
+   //
+   Real h = m_t_max / m_inital_row_length;
+
+   std::vector<Real> temp(m_inital_row_length + 1, Real(0));
+   for (std::size_t i = 0; i < m_inital_row_length; ++i)
+   {
+      Real t = h * i;
+      if ((first_complement < 0) && (t < m_t_crossover))
```

Comment:
Ugh... that logic is all wrong.... the "problem" is that if there's a bug in the first row's values the error generated gradually gets erased by subsequent rows so I didn't spot it.  Will fix by using the same logic used by subsequent rows (which are thankfully correct!)